File: hacker-rank/30-days-of-code/12-inheritance/main.rb
class Person
  def initialize(firstName, lastName, id)
    @firstName = firstName
    @lastName = lastName
    @id = id
  end
  def printPerson()
    print("Name: ",@lastName , ", " + @firstName ,"\nID: " , @id)
  end
end

class Student < Person
    def initialize(firstName, lastName, id, scores)
      @firstName = firstName
      @lastName = lastName
      @id = id
      @scores = scores.map { |v| Integer(v) }
    end

    def calculate
      average = @scores.inject(0) { |total, score| total + score } / @scores.size
      if average >= 90 && average <= 100
        'O'
      elsif average >= 80 && average < 90
        'E'
      elsif average >= 70 && average < 80
        'A'
      elsif average >= 55 && average < 70
        'P'
      elsif average >= 40 && average < 55
        'D'
      else
        'T'
      end
    end
end

input = gets.split()
firstName = input[0]
lastName = input[1]
id = input[2].to_i
numScores = gets.to_i
scores = gets.strip().split().map!(&:to_i)
s = Student.new(firstName, lastName, id, scores)
s.printPerson
print("\nGrade: " + s.calculate)

File: hacker-rank/30-days-of-code/12-inheritance/main.swift
// Class Person
class Person {
    var firstName: String
    var lastName: String
    var id: Int

    // Initializer
    init(firstName: String, lastName: String, identification: Int) {
        self.firstName = firstName
        self.lastName = lastName
        self.id = identification
    }

    // Print person data
    func printPerson() {
        print("Name: \(self.lastName), \(self.firstName)")
        print("ID: \(self.id)")
    }
} // End of class Person

// Class Student
class Student: Person {
    var testScores: [Int]

    init(firstName: String, lastName: String, identification: Int, scores: [Int]) {
        self.testScores = scores
        super.init(firstName: firstName, lastName: lastName, identification: identification)
    }

    func calculate() -> String {
        let average = self.testScores.reduce(0) { (t, score) -> Int in
            return t + score
        } / self.testScores.count
        if average >= 90 && average <= 100 {
          return "O"
        }
        else if average >= 80 && average < 90 {
          return "E"
        }
        else if (average >= 70 && average < 80) {
          return "A"
        }
        else if average >= 55 && average < 70 {
          return "P"
        }
        else if average >= 40 && average < 55 {
          return "D"
        }
        else {
          return "T"
        }
    }
} // End of class Student

let nameAndID = readLine()!.characters.split(" ").map{String($0)}
let scoreCount = readLine()
let scores = readLine()!.characters.split(" ").map{Int(String($0))!}

let s = Student(firstName: nameAndID[0], lastName: nameAndID[1], identification: Int(nameAndID[2])!, scores: scores)

s.printPerson()

print("Grade: \(s.calculate())")
